finality-aleph/src/network/manager/compatibility.rs

```diff
@@ -0,0 +1,153 @@
+        let num_bytes = ByteCount::decode(input)?;
+        match version {
+            0 => Ok(Legacy(NetworkData::decode(input)?)),
+            1 => Ok(V1(DiscoveryMessage::decode(input)?)),
+            _ => {
+                let mut payload = vec![
+                    0;
+                    num_bytes
+                        .try_into()
+                        .map_err(|_| "input too big to decode")?
+                ];
+                input.read(payload.as_mut_slice())?;
+                Ok(Other(version, payload))
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Error {
+    BadFormat,
+    UnknownVersion(Version),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            BadFormat => write!(f, "malformed encoding"),
+            UnknownVersion(version) => {
+                write!(f, "network data encoded with unknown version {}", version)
+            }
+        }
+    }
+}
+
+fn decode_pre_compatibility_network_data<D: Data, M: Multiaddress, LM: Multiaddress>(
+    data_raw: Vec<u8>,
+) -> Result<VersionedNetworkData<D, M, LM>, Error> {
+    use Error::*;
+
+    // We still have to be able to decode the pre-compatibility network data, so that
+    // we can complete a rolling update. When it becomes obsolete we can remove it.
+    match NetworkData::decode_all(&mut data_raw.as_slice()) {
+        Ok(data) => Ok(VersionedNetworkData::Legacy(data)),
+        Err(_) => Err(BadFormat),
+    }
+}
+
+/// Decodes Network Data, even if it was produced by ancient code which does not conform to our
+/// backwards compatibility style. This is reverse a funtion to `backwards_compatible_encode`.
+/// This should be removed, after rolling update with new network is completed
+pub fn backwards_compatible_decode<D: Data, M: Multiaddress, LM: Multiaddress>(
+    data_raw: Vec<u8>,
+) -> Result<VersionedNetworkData<D, M, LM>, Error> {
+    use Error::*;
+    let data_cloned = data_raw.clone();
+    match VersionedNetworkData::<D, M, LM>::decode_all(&mut data_cloned.as_slice()) {
+        Ok(data) => {
+            use VersionedNetworkData::*;
+            match data {
+                Legacy(data) => Ok(Legacy(data)),
+                V1(data) => Ok(V1(data)),
+                Other(version, _) => {
+                    // it is a coincidence that sometimes pre-compatibility legacy network data second word,
+                    // which is in VersionedNetworkData byte_count_size, can be small enough
+                    // so that network data is false positively recognized  as from the future
+                    // therefore we should try to decode formats
+                    decode_pre_compatibility_network_data(data_raw)
+                        .map_err(|_| UnknownVersion(version))
+                }
+            }
+        }
+        Err(_) => decode_pre_compatibility_network_data(data_raw),
+    }
+}
```

Comment:
We should write roundtrip tests for `Encode`/`Decode` instances of `VersionedNetworkData` -- i.e. make sure that the (de)serialization logic matches each other.

Reply:
done